common/src/main/java/io/pravega/common/io/filesystem/FileModificationEventWatcher.java

```diff
@@ -0,0 +1,225 @@
+ *   {@link FileModificationPollingMonitor} instead.
+ *
+ */
+@Slf4j
+public class FileModificationEventWatcher extends Thread implements FileModificationMonitor {
+
+    private static final AtomicInteger THREAD_NUM = new AtomicInteger();
+
+    /**
+     * The path of file to watch.
+     */
+    private final Path pathOfFileToWatch;
+
+    /**
+     * The action to perform when the specified file changes.
+     */
+    private final Consumer<WatchEvent<?>> callback;
+
+    private final UncaughtExceptionHandler uncaughtExceptionalHandler = (t, e) -> logException(e);
+
+    private boolean loopContinuously;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch the file to watch
+     * @param callback    the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @throws NullPointerException if either {@code fileToWatch} or {@code callback} is null
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     */
+    public FileModificationEventWatcher(@NonNull String fileToWatch, @NonNull Consumer<WatchEvent<?>> callback)
+            throws FileNotFoundException {
+        this(Paths.get(fileToWatch), callback, true);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param fileToWatch path of the file to watch
+     * @param callback          the callback to invoke when a modification to the {@code fileToWatch} is detected
+     * @param loopContinuously  whether to keep the thread look for file modification after one iteration. This option
+     *                          is useful for testing only.
+     * @throws InvalidPathException if {@code fileToWatch} is invalid
+     * @throws FileNotFoundException when a file at specified path {@code fileToWatch} does not exist
+     * @throws NullPointerException if either {@code fileToWatch}  or {@code callback} is null
+     */
+    @VisibleForTesting
+    FileModificationEventWatcher(Path fileToWatch, Consumer<WatchEvent<?>> callback,
+                                 boolean loopContinuously)
+            throws FileNotFoundException {
+        super();
+        this.validateInput(fileToWatch, callback);
+
+        this.pathOfFileToWatch = fileToWatch;
+        this.callback = callback;
+        setUncaughtExceptionHandler(uncaughtExceptionalHandler);
+
+        // Set the name for this object/thread for identification purposes.
+        this.setName("file-update-watcher-" + THREAD_NUM.incrementAndGet());
+        this.loopContinuously = loopContinuously;
+    }
+
+    @VisibleForTesting
+    String nameOfFileToWatch() {
+        Path fileName = this.pathOfFileToWatch.getFileName();
+        if (fileName != null) {
+            return fileName.toString();
+        } else {
+            throw new RuntimeException("File name is null");
```

Comment:
How can this possibly be null?
And if so, why throw the exception here? 
Furthermore, why throw `RuntimeException` when it would have thrown a `NullPointerException` anyway which would have been more informative.

Reply:
Spotbugs complains with a `NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE` (Possible null pointer dereference) error if I don't check for null here. I've modified the code to now throw `NullPointerException` now.